<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" lang="en"><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/><link rel="stylesheet" href="../jacoco-resources/report.css" type="text/css"/><link rel="shortcut icon" href="../jacoco-resources/report.gif" type="image/gif"/><title>StringEditDistance.java</title><link rel="stylesheet" href="../jacoco-resources/prettify.css" type="text/css"/><script type="text/javascript" src="../jacoco-resources/prettify.js"></script></head><body onload="window['PR_TAB_WIDTH']=4;prettyPrint()"><div class="breadcrumb" id="breadcrumb"><span class="info"><a href="../jacoco-sessions.html" class="el_session">Sessions</a></span><a href="../index.html" class="el_report">Jacoco Report</a> &gt; <a href="index.source.html" class="el_package">algorithms.alignment</a> &gt; <span class="el_source">StringEditDistance.java</span></div><h1>StringEditDistance.java</h1><pre class="source lang-java linenums">package algorithms.alignment;

import algorithms.util.PairIntArray;
import gnu.trove.list.TCharList;
import gnu.trove.list.array.TCharArrayList;
import gnu.trove.list.linked.TCharLinkedList;
import java.util.Arrays;

/**
 * find the number of insert, delete, and substitution operations to change
 * one string into another.

   note, may add a method for normalized edit distance in future:
     &quot;An Efficient Uniform-Cost Normalized Edit Distance Algorithm&quot;
      by Arslan and Egecioglu, 1999
      https://www.researchgate.net/publication/3820090_An_Efficient_Uniform-Cost_Normalized_Edit_Distance_Algorithm
 * 
 * NOTE: the alignment algorithms could be sped up to a runtime of O(N^2/log^2(N))
 * where N is max of string lengths, by using
 * Arlazarov, Dinic, Kronrod, Faradzev Four Russian algorithm.
 * 
 * NOTE also that the linear space algorithms can be improved by using
 * &quot;check point&quot; algorithms.  See Optimal alignments in linear space&quot;, Myers and Miller, 1988.
 * 
 first implemented in project
     https://github.com/nking/curvature-scale-space-corners-and-transformations
     w/ Copyright (c)Climb With Your Feet
     and using The MIT License (MIT)
   then moved to this shared library project which has the same copyright
   and license.
   
 * @author nichole
 */
<span class="fc" id="L34">public class StringEditDistance {</span>
   
    /**
     * calculate the number of insert, delete, and substitution operations to change
     * one string into another following Wagner and Fischer 1954 
     * &quot;The String-to-String Correction Problem&quot;.
     * This uses dynamic programming with time O(a.length * b.length)
     * and space O(a.length * b.length).
     * http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.367.5281&amp;rep=rep1&amp;type=pdf
     * 
     * Also, see Levenshtein Distance:
     * https://en.wikipedia.org/wiki/Levenshtein_distance
     * 
     * see also https://en.wikipedia.org/wiki/Needleman%E2%80%93Wunsch_algorithm
     * and
     * Vintsyuk TK (1968). &quot;Speech discrimination by dynamic programming&quot;. Kibernetika. 4: 81–88.
     * 
     * Other methods for comparing string similarity are Jaccard similarity, 
     * Jaro-Winkler distance, and dynamic time warping
     * (see thirdparty.fpetitjean.dba.DynamicTimeWarpingBarycenterAveraging in this project).
     * 
     * @param a string to edit
     * @param b target string to change a into.
     * 
     * @param outIndexes a and b indexes of the solution in pairs (a_i, b_j)
     * @return 
     */
    public int calculateWithWagnerFischer(String a, String b, PairIntArray outIndexes) {
        
<span class="fc" id="L63">        final int cDel = 1;</span>
<span class="fc" id="L64">        final int cIns = 1;</span>
         /*
        The positive costs and minimization are the common forms in this
        algorithm in context of computer science.
        In biology negative costs and maximization are often used instead
        to represent similarity.
        &quot;Reduced space sequence alignment&quot;
         Grice, Hughey, &amp; Speck, Bioinformatics, Volume 13, Issue 1, February 1997, 
         Pages 45–53, https://doi.org/10.1093/bioinformatics/13.1.45
        */
        
<span class="fc" id="L75">        int cChangeUnequal = 1;</span>
<span class="fc" id="L76">        int m = a.length();</span>
<span class="fc" id="L77">        int n = b.length();</span>
        // (m+1) X (n+1)
<span class="fc" id="L79">        int[][] d = new int[m + 1][];</span>
<span class="fc bfc" id="L80" title="All 2 branches covered.">        for (int i = 0; i &lt;= m; ++i) {</span>
<span class="fc" id="L81">            d[i] = new int[n + 1];</span>
        }
        
        //Theorem 2
        //D (i, j) is the cost of the least cost trace from A (i) to B (j).
        //D(i, j) = min( 
        //               D(i - 1, j - 1) + cost(A(i) -&gt; B(j)), 
        //               D(i- 1,j) + cost(A(i) deleted),
        //               D(i,j -- 1) + cost(insert a character that is  B(j}) )
        //          for all i, j, where 1 &lt;= i &lt;= [A|, 1 &lt;= j &lt;= [B|

        //Theorem 3
        //D(0, 0) = 0; D(i, 0) = summation_over_r=1_to_i(delete of A(r)); 
        //         and D(O, j) = summation_over_r=1_to_j(insert of a character that is B(r))
        // where 1 &lt;= i &lt;= [A|, 1 &lt;= j &lt;= [B|
        
        //Algorithm X
        //D[0,0] := 0;
        //for i := 1 to |A| do D[i, 0] := D[i - 1, 0] + cost_rm(A(i));
        //for j := 1 to |B| do D[O, j] := D[O,j - 1] + cost_insert(B(j));
        //for i := 1 to |A| do
        //  for j := 1 to |B| do begin
        //    m1 := D[i - 1, j - 1] + cost_change(A(i), B(j));
        //    m2 := D[i - 1, j] + cost_rm(A(i));
        //    m3 := D[i, j - 1] + cost_ins(B(j));
        //    D[i, j] := min(m1, m2, m3);
        //    end;
        
        /*
        NOTE: this algorithm is like the LongestCommonSubsequences except that
        for string edit: cIns=cDel=1 and cEqual=0, and the cost is minimized
        for LCS:         cIns=cDel=0 and cEqual=1, and the cost is maximized
        */
        
<span class="fc bfc" id="L115" title="All 2 branches covered.">        for (int i = 1; i &lt;= m; ++i) {</span>
<span class="fc" id="L116">            d[i][0] = i; //i*cDel</span>
        }
<span class="fc bfc" id="L118" title="All 2 branches covered.">        for (int i = 1; i &lt;= n; ++i) {</span>
            //a conversion of empty string 'a' into 'b' is all inserts
<span class="fc" id="L120">            d[0][i] = i; //i*cIns</span>
        }
        
        int i, j;
<span class="fc" id="L124">        int c = 0;</span>
        
<span class="fc bfc" id="L126" title="All 2 branches covered.">        for (j = 1; j &lt;= n; ++j) {</span>
<span class="fc bfc" id="L127" title="All 2 branches covered.">            for (i = 1; i &lt;= m; ++i) {</span>
<span class="fc bfc" id="L128" title="All 2 branches covered.">                if (a.charAt(i-1) == b.charAt(j-1)) {</span>
<span class="fc" id="L129">                    c = 0;</span>
                } else {
<span class="fc" id="L131">                    c = cChangeUnequal;</span>
                }
<span class="fc" id="L133">                d[i][j] = minimum(d[i-1][j-1] + c, d[i-1][j] + cDel, d[i][j-1] + cIns);</span>
            }
        }
        
        // Algorithm Y: print results
        // i := |A|;j := |B|;
        // while(i != 0 &amp;&amp; j != 0) do
        //    if D[i, j] = D[i - 1, j] + cost_rm(A(i)) then i := i - 1;
        //    else if D[i, j] = D[i,j - 1] + cost_insert(B(j)) then j :=j- 1;
        //    else begin
        //       print((i, j));
        //       i:=i- 1;j:=j-1;
        //       end;
<span class="fc" id="L146">        i = m;</span>
<span class="fc" id="L147">        j = n;</span>
<span class="pc bpc" id="L148" title="1 of 4 branches missed.">        while (i != 0 &amp;&amp; j != 0){</span>
<span class="fc bfc" id="L149" title="All 2 branches covered.">            if (d[i][j] == d[i-1][j] + cDel) {</span>
<span class="fc" id="L150">                i--;</span>
<span class="fc bfc" id="L151" title="All 2 branches covered.">            } else if (d[i][j] == d[i][j-1] + cIns) {</span>
<span class="fc" id="L152">                j--;</span>
            } else {
<span class="fc" id="L154">                outIndexes.add(i-1, j-1);// subtracting 1 for 0-based indexes</span>
<span class="fc" id="L155">                i--;</span>
<span class="fc" id="L156">                j--;</span>
            }
        }
        //System.out.printf(&quot;m=%d, n=%d, d=%n%s%n&quot;, m, n, FormatArray.toString(d, &quot;%d&quot;));
<span class="fc" id="L160">        return d[m][n];</span>
    }
    
    /**
     * calculate the number of insert, delete, and substitution operations to change
     * one string into another following Wagner and Fischer 1954 
     * &quot;The String-to-String Correction Problem&quot;, but modified to reduce the space complexity.
     * This uses dynamic programming with time O(a.length * b.length)
     * and space O(a.length).
     * &lt;pre&gt;
     * http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.367.5281&amp;rep=rep1&amp;type=pdf
     * 
     * for space complexity reduction, lecture slides of
     * CS 473: Fundamental Algorithms, Spring 2011
     https://courses.engr.illinois.edu/cs473/sp2011/Lectures/09_lec.pdf
     Slide Acknowledgments: Lecture notes are based on notes of Jeff Erickson, 
     Sariel Har-Peled, Chandra Chekuri and the slides of Mahesh Viswanathan and 
     the books by Kleinberg and Tardos, and by Dasgupta, Papadimitrioiu and Vazirani.

     * Also, see Levenshtein Distance:
     * https://en.wikipedia.org/wiki/Levenshtein_distance
     * &lt;/pre&gt;
     * @param a string to edit
     * @param b target string to change a into.
     * 
     * @return the edit distance which is the minimum number of deletes of 
     * characters in string a and inserts of characters into string b to turn
     * string a into string b.
     */
    public int calculateWithModifiedWagnerFischer(String a, String b) {
       
<span class="fc" id="L191">        final int cDel = 1;</span>
<span class="fc" id="L192">        final int cIns = 1;</span>
        /*
        The positive costs and minimization are the common forms in this
        algorithm in context of computer science.
        In biology negative costs and maximization are often used instead
        to represent similarity.
        &quot;Reduced space sequence alignment&quot;
         Grice, Hughey, &amp; Speck, Bioinformatics, Volume 13, Issue 1, February 1997, 
         Pages 45–53, https://doi.org/10.1093/bioinformatics/13.1.45
        */
       
        /*
            Entries in jth column only depend on (j − 1)’s column and earlier 
               entries in jth column.
            Only store the current column and the previous column reusing space; 
               N(i, 0) stores M(i, j − 1) and N(i, 1) stores M(i, j)
        */
        
<span class="fc" id="L210">        int cChangeUnequal = 1;</span>
<span class="fc" id="L211">        int m = a.length();</span>
<span class="fc" id="L212">        int n = b.length();</span>
<span class="fc" id="L213">        int[][] dN = new int[m + 1][];</span>
        int i;
        //dN(i, 0) stores M(i, j − 1) and dN(i, 1) stores M(i, j)
<span class="fc bfc" id="L216" title="All 2 branches covered.">        for (i = 0; i &lt;= m; ++i) {</span>
<span class="fc" id="L217">            dN[i] = new int[2];</span>
        }
        // (m+1) X (n+1)
        /*int[][] d = new int[m + 1][];
        for (int i = 0; i &lt;= m; ++i) {
            d[i] = new int[n + 1];
        }*/
        
        //Theorem 2
        //D (i, j) is the cost of the least cost trace from A (i) to B (j).
        //D(i, j) = min( 
        //               D(i - 1, j - 1) + cost(A(i) -&gt; B(j)), 
        //               D(i- 1,j) + cost(A(i) deleted),
        //               D(i,j -- 1) + cost(insert a character that is  B(j}) )
        //          for all i, j, where 1 &lt;= i &lt;= [A|, 1 &lt;= j &lt;= [B|

        //Theorem 3
        //D(0, 0) = 0; D(i, 0) = summation_over_r=1_to_i(delete of A(r)); 
        //         and D(O, j) = summation_over_r=1_to_j(insert of a character that is B(r))
        // where 1 &lt;= i &lt;= [A|, 1 &lt;= j &lt;= [B|
        
        //Algorithm X
        //D[0,0] := 0;
        //for i := 1 to |A| do D[i, 0] := D[i - 1, 0] + cost_rm(A(i));
        //for j := 1 to |B| do D[O, j] := D[O,j - 1] + cost_insert(B(j));
        //for i := 1 to |A| do
        //  for j := 1 to |B| do begin
        //    m1 := D[i - 1, j - 1] + cost_change(A(i), B(j));
        //    m2 := D[i - 1, j] + cost_rm(A(i));
        //    m3 := D[i, j - 1] + cost_ins(B(j));
        //    D[i, j] := min(m1, m2, m3);
        //    end;
        
        /*
        NOTE: this algorithm is like the LongestCommonSubsequences except that
        for string edit: cIns=cDel=1 and cEqual=0, and the cost is minimized
        for LCS:         cIns=cDel=0 and cEqual=1, and the cost is maximized
        */
        
        //dN(i, 0) stores M(i, j − 1) and dN(i, 1) stores M(i, j)
<span class="fc bfc" id="L257" title="All 2 branches covered.">        for (i = 1; i &lt;= m; ++i) {</span>
            //d[i][0] = i; //i*cDel
<span class="fc" id="L259">            dN[i][0] = i; //i*cDel</span>
        }
        //for (i = 1; i &lt;= n; ++i) {
            //d[0][i] = i; //i*cIns
        //}
        
        int j;
<span class="fc" id="L266">        int c = 0;</span>
        
        int minPos;
                
<span class="fc bfc" id="L270" title="All 2 branches covered.">        for (j = 1; j &lt;= n; ++j) {</span>
<span class="fc" id="L271">            dN[0][1] = j; //j*cIns</span>
            //System.out.printf(&quot;j=%d: &quot;, j-1);
<span class="fc bfc" id="L273" title="All 2 branches covered.">            for (i = 1; i &lt;= m; ++i) {</span>
<span class="fc bfc" id="L274" title="All 2 branches covered.">                if (a.charAt(i-1) == b.charAt(j-1)) {</span>
<span class="fc" id="L275">                    c = 0;</span>
                    //System.out.printf(&quot;(%d,%d),  &quot;, i-1, j-1); //subtracting 1 for 0-based indexes
                } else {
<span class="fc" id="L278">                    c = cChangeUnequal;</span>
                }
                //d[i][j] = minimum(d[i-1][j-1] + c, d[i-1][j] + cDel, d[i][j-1] + cIns);
<span class="fc" id="L281">                minPos = minimumPos(dN[i-1][0] + c, dN[i-1][1] + cDel, dN[i][0] + cIns);</span>
<span class="fc bfc" id="L282" title="All 3 branches covered.">                switch(minPos) {</span>
                    case 1: {
<span class="fc" id="L284">                        dN[i][1] = dN[i-1][1] + cDel;</span>
<span class="fc" id="L285">                        break;</span>
                    } case 2: {
<span class="fc" id="L287">                        dN[i][1] = dN[i][0] + cIns;</span>
<span class="fc" id="L288">                        break;</span>
                    } default: {
<span class="fc" id="L290">                        dN[i][1] = dN[i-1][0] + c; </span>
                        break;
                    }
                }
            }
            //System.out.printf(&quot;%n, &quot;);
            
            // for i = 1 to m do Copy N[i, 0] = N[i, 1]
<span class="fc bfc" id="L298" title="All 2 branches covered.">            for (i = 1; i &lt;= m; ++i) {</span>
<span class="fc" id="L299">                dN[i][0] = dN[i][1];</span>
            }
        }
         
        /*
        can use checkpoint methods to recover the alignment in O(n) space where
        n is the length of the longest string:
        
        &quot;Reduced space sequence alignment&quot;
         Grice, Hughey, &amp; Speck, Bioinformatics, Volume 13, Issue 1, February 1997, 
         Pages 45–53, https://doi.org/10.1093/bioinformatics/13.1.45
        
        The Ukkonen algorithm with the Powell checkpoint method can be used
        to recover alignment in addition to the edit distance using
        O(N) space:
        &quot;A versatile divide and conquer technique for optimal string alignment&quot;
         Powell, Allison, and Dix, Information Processing Letters 70 (1999) 127–139 
        
        alignment can be global, semi-lobal, local, optimal, approximate, 
        pairwise, mutlple sequence, etc.
        
        A 2004 review of use of it in biological context can be found the book 
        &quot;Bioinformatics Sequence and Genome
        Analysis&quot; by David Mount.
        */
        
<span class="fc" id="L325">        return dN[m][1];</span>
    }
    
        /**
     * calculate the number of insert, delete, and substitution operations to change
     * one string into another as an implementation of the Needleman-Wunsch
     * scoring. (The scoring is a modification of the modified Wagner-Fischer algorithm.)
     * &lt;pre&gt;
     * The pseudocode is at https://en.wikipedia.org/wiki/Hirschberg%27s_algorithm
     * see also
     * LongestCommonSubsequence.java
     * &lt;/pre&gt;
     * @param a string to edit
     * @param b target string to change a into.
     * 
     * @return the edit distance which is the minimum number of deletes of 
     * characters in string a and inserts of characters into string b to turn
     * string a into string b.
     */
    public int calculateNeedlemanWunschScoring(char[] a, char[] b) {
       
<span class="fc" id="L346">        final int cDel = -2;</span>
<span class="fc" id="L347">        final int cIns = -2;</span>
<span class="fc" id="L348">        final int subUnequal = -1;</span>
<span class="fc" id="L349">        final int subEqual = 2;</span>
       
        /*
            Entries in jth column only depend on (j − 1)’s column and earlier 
               entries in jth column.
            Only store the current column and the previous column reusing space; 
               N(i, 0) stores M(i, j − 1) and N(i, 1) stores M(i, j)
        */
        
<span class="fc" id="L358">        int m = a.length;</span>
<span class="fc" id="L359">        int n = b.length;</span>
<span class="fc" id="L360">        int[][] dN = new int[m + 1][];</span>
        int i;
        //dN(i, 0) stores M(i, j − 1) and dN(i, 1) stores M(i, j)
<span class="fc bfc" id="L363" title="All 2 branches covered.">        for (i = 0; i &lt;= m; ++i) {</span>
<span class="fc" id="L364">            dN[i] = new int[2];</span>
        }
<span class="fc bfc" id="L366" title="All 2 branches covered.">        for (i = 1; i &lt;= m; ++i) {</span>
<span class="fc" id="L367">            dN[i][0] = i*cDel;</span>
        }
        
        int j;
<span class="fc" id="L371">        int c = 0;</span>
                        
<span class="fc bfc" id="L373" title="All 2 branches covered.">        for (j = 1; j &lt;= n; ++j) {</span>
<span class="fc" id="L374">            dN[0][1] = j*cIns;</span>
<span class="fc bfc" id="L375" title="All 2 branches covered.">            for (i = 1; i &lt;= m; ++i) {</span>
<span class="fc bfc" id="L376" title="All 2 branches covered.">                if (a[i-1] == b[j-1]) {</span>
<span class="fc" id="L377">                    c = subEqual;</span>
                    //System.out.printf(&quot;(%d,%d),  &quot;, i-1, j-1); //subtracting 1 for 0-based indexes
                } else {
<span class="fc" id="L380">                    c = subUnequal;</span>
                }
<span class="fc" id="L382">                dN[i][1] = maximum(dN[i-1][0] + c, dN[i-1][1] + cDel, dN[i][0] + cIns);</span>
            }
            //System.out.printf(&quot;%n, &quot;);
            
            // for i = 1 to m do Copy N[i, 0] = N[i, 1]
<span class="fc bfc" id="L387" title="All 2 branches covered.">            for (i = 1; i &lt;= m; ++i) {</span>
<span class="fc" id="L388">                dN[i][0] = dN[i][1];</span>
            }
        }
        
<span class="fc" id="L392">        return dN[m][1];</span>
    }
    
    /**
     * find the optimal string metric distance between 2 strings using
     * Hirschberg's algorithm.  The scoring uses Levenshtein distance, 
     * defined to be the sum of the costs of insertions, replacements, deletions, 
     * and null actions needed to change one string into the other. 
     * &lt;pre&gt;
     * https://en.wikipedia.org/wiki/Hirschberg%27s_algorithm
     * &lt;/pre&gt;
     * @param a
     * @param b
     * @return 
     */
    public TCharList[] hirschbergOptimal(char[] a, char[] b) {
                
<span class="fc" id="L409">        TCharList z = new TCharArrayList();</span>
<span class="fc" id="L410">        TCharList w = new TCharLinkedList();</span>
        
<span class="fc" id="L412">        int m = a.length;</span>
<span class="fc" id="L413">        int n = b.length;</span>
        int i;
<span class="pc bpc" id="L415" title="1 of 2 branches missed.">        if (m == 0) {</span>
<span class="nc bnc" id="L416" title="All 2 branches missed.">            for (i = 0; i &lt; n; ++i) {</span>
<span class="nc" id="L417">                z.add('-');</span>
<span class="nc" id="L418">                w.add(b[i]);</span>
            }
<span class="nc" id="L420">            return new TCharList[]{z, w};</span>
<span class="fc bfc" id="L421" title="All 2 branches covered.">        } else if (n == 0) {</span>
<span class="fc bfc" id="L422" title="All 2 branches covered.">            for (i = 0; i &lt; m; ++i) {</span>
<span class="fc" id="L423">                z.add(a[i]);</span>
<span class="fc" id="L424">                w.add('-');</span>
            }
<span class="fc" id="L426">            return new TCharList[]{z, w};</span>
<span class="pc bpc" id="L427" title="1 of 4 branches missed.">        } else if (m==1 || n==1) {</span>
<span class="fc" id="L428">            PairIntArray outIndexes = new PairIntArray();</span>
<span class="fc" id="L429">            int nEdits = calculateWithWagnerFischer(new String(a), new String(b), outIndexes);</span>
<span class="fc" id="L430">            char[] resultA = new char[outIndexes.getN()];</span>
<span class="fc" id="L431">            char[] resultB = new char[outIndexes.getN()];</span>
<span class="fc bfc" id="L432" title="All 2 branches covered.">            for (int ii = 0; ii &lt; resultA.length; ++ii) {</span>
<span class="fc" id="L433">                resultA[ii] = a[outIndexes.getX(ii)];</span>
<span class="fc" id="L434">                resultB[ii] = b[outIndexes.getY(ii)];</span>
            }
<span class="fc" id="L436">            z.addAll(resultA);</span>
<span class="fc" id="L437">            w.addAll(resultB);</span>
<span class="fc" id="L438">            return new TCharList[]{z, w};</span>
        }
        //m: xlen = length(X)
<span class="fc" id="L441">        int aMid = m/2;</span>
        //n: ylen = length(Y)
        //ScoreL = NWScore(X1:xmid, Y)
        //int scoreL = calculateNeedlemanWunschScoring(Arrays.copyOfRange(a, 0, aMid), b);
        //ScoreR = NWScore(rev(Xxmid+1:xlen), rev(Y))
        //int scoreR = calculateNeedlemanWunschScoring(
        //    reverse2(Arrays.copyOfRange(a, aMid + 1, m)), reverse2(b));
        //ymid = arg max ScoreL + rev(ScoreR)
        /* &quot;Optimal alignments in linear space&quot;, Myers and Miller, 1988
        Let rev(A) denote the reverse of A, i.e. a_{M-1}a_M...a_0
        and A^T denote the suffix a_{i+1}a_{i+2}...a_M of A. 
        http://www.cs.tau.ac.il/~rshamir/algmb/98/scribe/html/lec02/node9.html
        find k in 0&lt;=k&lt;=n that maximizes D(m,n) = {D(m/2, k) + D^T(m/2, n-k) }
            where 
            D(m/2, k) is D(string1[1:m/2], string2[1:k])
            = D( a.substring(0, aMid), b.substring(0, k) )
            and D^T(m/2, n-k) is 
              D( reversed string1[(m/2)+1:m], reversed string2[k+1:n])
            = D( reverse( a.substring(aMid + 1, m)),reverse( b.substring(k + 1, n)) )
        bMid = ymid = k;
        */
<span class="fc" id="L462">        int bMid = argmax(aMid, a, b);</span>
                
        //(Z,W) = Hirschberg(X1:xmid, y1:ymid) + Hirschberg(Xxmid+1:xlen, Yymid+1:ylen)
        
<span class="fc" id="L466">        TCharList[] zW0 = hirschbergOptimal(</span>
<span class="fc" id="L467">            Arrays.copyOfRange(a, 0, aMid), Arrays.copyOfRange(b, 0, bMid));</span>
<span class="fc" id="L468">        TCharList[] zW1 = hirschbergOptimal( </span>
<span class="fc" id="L469">            Arrays.copyOfRange(a, aMid, m), Arrays.copyOfRange(b, bMid, n));</span>
        
<span class="fc" id="L471">        TCharList[] zW = new TCharList[2];</span>
<span class="fc" id="L472">        zW[0] = new TCharArrayList(zW0[0]);</span>
<span class="fc" id="L473">        zW[0].addAll(zW1[0]);</span>
<span class="fc" id="L474">        zW[1] = new TCharArrayList(zW0[1]);</span>
<span class="fc" id="L475">        zW[1].addAll(zW1[1]);</span>
        
<span class="fc" id="L477">        return zW;</span>
    }
    
    private String reverse(String a) {
<span class="nc" id="L481">        char[] r = a.toCharArray();</span>
<span class="nc" id="L482">        reverse(r);</span>
<span class="nc" id="L483">        return new String(r);</span>
    }
    private void reverse(char[] a) {
<span class="fc" id="L486">        int n = a.length;</span>
<span class="fc" id="L487">        int i, i2, mid = n/2;</span>
        char swap;
<span class="fc bfc" id="L489" title="All 2 branches covered.">        for (i = 0, i2=n-1; i &lt; mid; ++i, i2--) {</span>
<span class="fc" id="L490">            swap = a[i];</span>
<span class="fc" id="L491">            a[i] = a[i2];</span>
<span class="fc" id="L492">            a[i2] = swap;</span>
        }
<span class="fc" id="L494">    }</span>
    private char[] reverse2(char[] a) {
<span class="nc" id="L496">        char[] out = Arrays.copyOf(a, a.length);</span>
<span class="nc" id="L497">        reverse(out);</span>
<span class="nc" id="L498">        return out;</span>
    }
    
    private int minimum(int a, int b, int c) {
<span class="fc bfc" id="L502" title="All 2 branches covered.">        if (a &lt;= b) {</span>
<span class="fc bfc" id="L503" title="All 2 branches covered.">            if (a &lt;= c) {</span>
<span class="fc" id="L504">                return a;</span>
            } else {
<span class="fc" id="L506">                return c;</span>
            }
        } else {
            // b &lt; a
<span class="pc bpc" id="L510" title="1 of 2 branches missed.">            if (b &lt;= c) {</span>
<span class="fc" id="L511">                return b;</span>
            } else {
<span class="nc" id="L513">                return c;</span>
            }
        } 
    }
    private int minimumPos(int a, int b, int c) {
<span class="fc bfc" id="L518" title="All 2 branches covered.">        if (a &lt;= b) {</span>
<span class="fc bfc" id="L519" title="All 2 branches covered.">            if (a &lt;= c) {</span>
<span class="fc" id="L520">                return 0;</span>
            } else {
<span class="fc" id="L522">                return 2;</span>
            }
        } else {
            // b &lt; a
<span class="pc bpc" id="L526" title="1 of 2 branches missed.">            if (b &lt;= c) {</span>
<span class="fc" id="L527">                return 1;</span>
            } else {
<span class="nc" id="L529">                return 2;</span>
            }
        } 
    }
    
    private int maximum(int a, int b, int c) {
<span class="fc bfc" id="L535" title="All 2 branches covered.">        if (a &gt;= b) {</span>
<span class="fc bfc" id="L536" title="All 2 branches covered.">            if (a &gt;= c) {</span>
<span class="fc" id="L537">                return a;</span>
            } else {
<span class="fc" id="L539">                return c;</span>
            }
        } else {
            // b &gt; a
<span class="pc bpc" id="L543" title="1 of 2 branches missed.">            if (b &gt;= c) {</span>
<span class="fc" id="L544">                return b;</span>
            } else {
<span class="nc" id="L546">                return c;</span>
            }
        } 
    }

    /**
     * find the index in b which maximizes the total score
     *    D( a.substring(0, aMid), b.substring(0, k) )
             + D( reverse( a.substring(aMid + 1, m)), reverse(b.substring(k + 1, n)) ) );
       where D is the score from algorithm NeedlemanWunsch.
     * @param aMid
     * @param a
     * @param b
     * @return 
     */
    private int argmax(int aMid, char[] a, char[] b) {
        /*
        D( a.substring(0, aMid), b.substring(0, k) )
             + D( reverse( a.substring(aMid + 1, m) ), 
                  reverse( b.substring(k + 1, n) ) );
        */
<span class="fc" id="L567">        char[] a0Mid = Arrays.copyOfRange(a, 0, aMid);</span>
<span class="fc" id="L568">        char[] aMidMRev = Arrays.copyOfRange(a, aMid, a.length);</span>
<span class="fc" id="L569">        reverse(aMidMRev);</span>
        
<span class="fc" id="L571">        int maxK = -1;</span>
<span class="fc" id="L572">        int maxVal = Integer.MIN_VALUE;</span>
        int k, sL, sR, tot;
        char[] b0k, bkMidRev;
<span class="fc bfc" id="L575" title="All 2 branches covered.">        for (k = 0; k &lt; b.length; ++k) {</span>
<span class="fc" id="L576">            b0k = Arrays.copyOfRange(b, 0, k);</span>
<span class="fc" id="L577">            bkMidRev = Arrays.copyOfRange(b, k, b.length);</span>
<span class="fc" id="L578">            reverse(bkMidRev);</span>
            
<span class="fc" id="L580">            sL = calculateNeedlemanWunschScoring(a0Mid, b0k);</span>
<span class="fc" id="L581">            sR = calculateNeedlemanWunschScoring(aMidMRev, bkMidRev);</span>
            
<span class="fc" id="L583">            tot = sL + sR;</span>
<span class="fc bfc" id="L584" title="All 2 branches covered.">            if (tot &gt; maxVal) {</span>
<span class="fc" id="L585">                maxVal = tot;</span>
<span class="fc" id="L586">                maxK = k;</span>
            }
        }
<span class="fc" id="L589">        return maxK;</span>
    }
}
</pre><div class="footer"><span class="right">Created with <a href="http://www.jacoco.org/jacoco">JaCoCo</a> 0.8.8.202204050719</span></div></body></html>